File: src/claudeline/generate_cert.py
"""Self-signed certificate generation for HTTPS support.

Run as: uv run python -m claudeline.generate_cert

Generates a self-signed TLS certificate suitable for local network use,
enabling HTTPS so that mobile browsers allow microphone access (getUserMedia).
"""

import datetime
import ipaddress
import os
import socket
import sys
from pathlib import Path

from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.x509.oid import NameOID


def get_local_ips() -> list[str]:
    """Detect local IPv4 addresses (excluding loopback).

    Uses socket.getaddrinfo() for cross-platform compatibility.
    Falls back to the UDP socket trick if no addresses are found.
    """
    ips: set[str] = set()

    # Primary method: getaddrinfo on hostname
    try:
        hostname = socket.gethostname()
        for info in socket.getaddrinfo(hostname, None, socket.AF_INET):
            addr = str(info[4][0])
            if not addr.startswith("127."):
                ips.add(addr)
    except OSError:
        pass

    # Fallback: UDP socket trick
    if not ips:
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
                s.connect(("8.8.8.8", 80))
                addr = s.getsockname()[0]
                if not addr.startswith("127."):
                    ips.add(addr)
        except OSError:
            pass

    return sorted(ips)


def generate_certificate(
    ips: list[str],
    output_dir: str = "certs",
    days_valid: int = 365,
    common_name: str = "Claude Line Local",
) -> tuple[Path, Path]:
    """Generate a self-signed EC P-256 certificate with SANs for local IPs.

    Args:
        ips: Additional IP addresses to include as SANs.
        output_dir: Directory to write cert.pem and key.pem into.
        days_valid: Certificate validity period in days.
        common_name: CN for the certificate subject/issuer.

    Returns:
        Tuple of (cert_path, key_path) as absolute Paths.
    """
    # Generate EC P-256 private key
    key = ec.generate_private_key(ec.SECP256R1())

    # Build subject/issuer
    name = x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, common_name)])

    # Build SANs: always include localhost + 127.0.0.1, plus user-provided IPs
    sans: list[x509.GeneralName] = [
        x509.DNSName("localhost"),
        x509.IPAddress(ipaddress.IPv4Address("127.0.0.1")),
    ]
    for ip_str in ips:
        try:
            ip_addr = ipaddress.IPv4Address(ip_str)
            san = x509.IPAddress(ip_addr)
            if san not in sans:
                sans.append(san)
        except ValueError:
            continue

    now = datetime.datetime.now(datetime.UTC)
    cert = (
        x509.CertificateBuilder()
        .subject_name(name)
        .issuer_name(name)
        .public_key(key.public_key())
        .serial_number(x509.random_serial_number())
        .not_valid_before(now)
        .not_valid_after(now + datetime.timedelta(days=days_valid))
        .add_extension(
            x509.SubjectAlternativeName(sans),
            critical=False,
        )
        .add_extension(
            x509.BasicConstraints(ca=True, path_length=0),
            critical=True,
        )
        .sign(key, hashes.SHA256())
    )

    # Write files
    out = Path(output_dir)
    out.mkdir(parents=True, exist_ok=True)

    key_path = out / "key.pem"
    cert_path = out / "cert.pem"

    key_path.write_bytes(
        key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption(),
        )
    )
    if os.name != "nt":
        key_path.chmod(0o600)

    cert_path.write_bytes(cert.public_bytes(serialization.Encoding.PEM))

    return cert_path.resolve(), key_path.resolve()


def print_instructions(cert_path: Path, key_path: Path, ips: list[str]) -> None:
    """Print setup instructions after certificate generation."""
    port = 8765

    print("\n--- Certificate Generated ---")
    print(f"  Certificate: {cert_path}")
    print(f"  Private key: {key_path}")

    print("\n--- Start Server with HTTPS ---")
    print(
        f"  claude-line --ssl-certfile {cert_path} --ssl-keyfile {key_path}"
    )

    if ips:
        print("\n--- Access from your phone ---")
        for ip in ips:
            print(f"  https://{ip}:{port}")

    print("\n--- iOS: Trust the certificate ---")
    print("  1. Transfer cert.pem to your iPhone (AirDrop, email, or HTTP)")
    print("  2. Open it to install the profile")
    print("  3. Go to Settings > General > About > Certificate Trust Settings")
    print("  4. Enable full trust for 'Claude Line Local'")

    print("\n--- Android: Trust the certificate ---")
    print("  1. Transfer cert.pem to your Android device")
    print("  2. Go to Settings > Security > Install certificate > CA certificate")
    print("  3. Select the cert.pem file")

    print("\n--- Browser: Accept the warning ---")
    print("  Navigate to the HTTPS URL and accept the self-signed cert warning.")
    print("  (Chrome: 'Advanced' > 'Proceed'; Firefox: 'Accept the Risk')")
    print()


def interactive_main() -> None:
    """Interactive certificate generation entry point."""
    print("Claude Line — Self-Signed Certificate Generator\n")

    detected_ips = get_local_ips()

    if detected_ips:
        print("Detected local IP addresses:")
        for ip in detected_ips:
            print(f"  {ip}")
    else:
        print("No local IP addresses detected.")

    selected_ips = detected_ips

    # Interactive mode: let user confirm/add IPs
    if sys.stdin.isatty():
        print(
            "\nPress Enter to use all detected IPs, or type IPs "
            "separated by spaces:"
        )
        try:
            user_input = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            return

        if user_input:
            selected_ips = user_input.split()

    print(f"\nGenerating certificate for IPs: {', '.join(selected_ips) or 'none'}")
    cert_path, key_path = generate_certificate(selected_ips)
    print_instructions(cert_path, key_path, selected_ips)


if __name__ == "__main__":
    interactive_main()

File: src/generate-cert.js
/**
 * Self-signed certificate generation for HTTPS support.
 *
 * Run as: node src/generate-cert.js
 *
 * Generates a self-signed TLS certificate suitable for local network use,
 * enabling HTTPS so that mobile browsers allow microphone access (getUserMedia).
 */

import { mkdirSync, writeFileSync, chmodSync } from 'fs';
import { createInterface } from 'readline';
import { networkInterfaces } from 'os';
import { resolve } from 'path';
import { fileURLToPath } from 'url';

import selfsigned from 'selfsigned';

/**
 * Detect local IPv4 addresses (excluding loopback).
 *
 * @returns {string[]} Sorted list of local IPv4 addresses.
 */
export function getLocalIps() {
  const ips = new Set();
  const ifaces = networkInterfaces();

  for (const name of Object.keys(ifaces)) {
    for (const iface of ifaces[name]) {
      if (iface.family === 'IPv4' && !iface.internal) {
        ips.add(iface.address);
      }
    }
  }

  return [...ips].sort();
}

/**
 * Generate a self-signed RSA certificate with SANs for local IPs.
 *
 * @param {string[]} ips - IP addresses to include as SANs.
 * @param {object} [options]
 * @param {string} [options.outputDir='certs'] - Directory for cert.pem and key.pem.
 * @param {number} [options.daysValid=365] - Certificate validity period.
 * @param {string} [options.commonName='Claude Line Local'] - CN for the certificate.
 * @returns {{ certPath: string, keyPath: string }} Absolute paths to generated files.
 */
export async function generateCertificate(
  ips,
  { outputDir = 'certs', daysValid = 365, commonName = 'Claude Line Local' } = {},
) {
  // Build SAN entries: always include localhost + 127.0.0.1
  const altNames = [
    { type: 2, value: 'localhost' }, // DNS
    { type: 7, ip: '127.0.0.1' }, // IP
  ];

  const seenIps = new Set(['127.0.0.1']);
  for (const ip of ips) {
    if (!seenIps.has(ip) && /^\d{1,3}(\.\d{1,3}){3}$/.test(ip)) {
      altNames.push({ type: 7, ip });
      seenIps.add(ip);
    }
  }

  const attrs = [{ name: 'commonName', value: commonName }];

  const pems = await selfsigned.generate(attrs, {
    keySize: 2048, // RSA key size (selfsigned uses RSA)
    days: daysValid,
    algorithm: 'sha256',
    extensions: [
      {
        name: 'basicConstraints',
        cA: true,
        pathLenConstraint: 0,
        critical: true,
      },
      {
        name: 'subjectAltName',
        altNames,
      },
    ],
  });

  // Write files
  mkdirSync(outputDir, { recursive: true });

  const certPath = resolve(outputDir, 'cert.pem');
  const keyPath = resolve(outputDir, 'key.pem');

  writeFileSync(certPath, pems.cert);
  writeFileSync(keyPath, pems.private);

  // Restrict key file permissions on Unix
  if (process.platform !== 'win32') {
    try {
      chmodSync(keyPath, 0o600);
    } catch {
      // Ignore permission errors
    }
  }

  return { certPath, keyPath };
}

/**
 * Print setup instructions after certificate generation.
 *
 * @param {string} certPath
 * @param {string} keyPath
 * @param {string[]} ips
 */
export function printInstructions(certPath, keyPath, ips) {
  const port = 8765;

  console.log('\n--- Certificate Generated ---');
  console.log(`  Certificate: ${certPath}`);
  console.log(`  Private key: ${keyPath}`);

  console.log('\n--- Start Server with HTTPS ---');
  console.log(`  SSL_CERTFILE=${certPath} SSL_KEYFILE=${keyPath} npm start`);

  if (ips.length > 0) {
    console.log('\n--- Access from your phone ---');
    for (const ip of ips) {
      console.log(`  https://${ip}:${port}`);
    }
  }

  console.log('\n--- iOS: Trust the certificate ---');
  console.log('  1. Transfer cert.pem to your iPhone (AirDrop, email, or HTTP)');
  console.log('  2. Open it to install the profile');
  console.log('  3. Go to Settings > General > About > Certificate Trust Settings');
  console.log("  4. Enable full trust for 'Claude Line Local'");

  console.log('\n--- Android: Trust the certificate ---');
  console.log('  1. Transfer cert.pem to your Android device');
  console.log('  2. Go to Settings > Security > Install certificate > CA certificate');
  console.log('  3. Select the cert.pem file');

  console.log('\n--- Browser: Accept the warning ---');
  console.log('  Navigate to the HTTPS URL and accept the self-signed cert warning.');
  console.log("  (Chrome: 'Advanced' > 'Proceed'; Firefox: 'Accept the Risk')");
  console.log();
}

/**
 * Interactive certificate generation entry point.
 */
export async function interactiveMain() {
  console.log('Claude Line — Self-Signed Certificate Generator\n');

  const detectedIps = getLocalIps();

  if (detectedIps.length > 0) {
    console.log('Detected local IP addresses:');
    for (const ip of detectedIps) {
      console.log(`  ${ip}`);
    }
  } else {
    console.log('No local IP addresses detected.');
  }

  let selectedIps = detectedIps;

  // Interactive mode: let user confirm/add IPs
  if (process.stdin.isTTY) {
    const rl = createInterface({ input: process.stdin, output: process.stdout });
    const answer = await new Promise((resolve) => {
      rl.question(
        '\nPress Enter to use all detected IPs, or type IPs separated by spaces:\n> ',
        resolve,
      );
    });
    rl.close();

    const trimmed = answer.trim();
    if (trimmed) {
      selectedIps = trimmed.split(/\s+/);
    }
  }

  console.log(`\nGenerating certificate for IPs: ${selectedIps.join(', ') || 'none'}`);
  const { certPath, keyPath } = await generateCertificate(selectedIps);
  printInstructions(certPath, keyPath, selectedIps);
}

// Run if executed directly
const isMain =
  process.argv[1] && resolve(process.argv[1]) === resolve(fileURLToPath(import.meta.url));
if (isMain) {
  interactiveMain();
}
